### ISSUE TITLE
Low d with Flux model, general feedback

### ISSUE THREAD

**Issue by rockerBOO** [LoganBooker/prodigy-plus-schedule-free#14]
Hello! Thanks for this optimizer, I have been telling everyone about it. 

I have gotten great results using it with SD1.5 and SDXL LoRA training, exceeding my previous runs. It supporting parameter groups allows it to work with LoRA+ which trains the LoRA network of B and A to increase the LR of B. This has worked really well on these models.

I have been recently trying it on Flux.dev, flux.dev2pro, and flex.1 alpha model for LoRA training. But I am finding the d ends up being really small compared to what results I would expect. 

Example chart showing Prodigy d*lr on the bottom. And comparing it with AdamW 5e-4 run. 
![Image](https://github.com/user-attachments/assets/ae5ed417-427b-4c27-b4bd-364f5662c3af)

Example showing d*lr for Flux LoRA
![Image](https://github.com/user-attachments/assets/3089ab94-5c44-447b-933a-42d407fd3ad9)

I've tried all sorts of things to improve the performance but might be difficult to outline here. 

Minimally using Kohya-ss trainer
```
optimizer_type = "prodigyplus.ProdigyPlusScheduleFree"
optimizer_args = [
  "eps=None",
  "use_speed=True",
  "use_adopt=True",
]
```

I've also tried combos of these values at different ranges but getting very minimal changes to the end results.
```
  "weight_decay=5",
  "d_coef=4",
  "use_bias_correction=True",
  "use_cautious=True",
  "use_orthograd=True",
  "split_groups_mean=False",
```

I'm also using LoRA+ but I haven't done it without LoRA+ https://arxiv.org/abs/2402.12354

```
  "loraplus_unet_lr_ratio=16",
```

This is mixed with the difficulty of flux dev training, but similar results on flux1 dev2pro, and flex1 alpha. 

You mentioned that it would be good to keep track of the norms of the weights so I made some adjustments to record these. Comparing AdamW 5e-5 (LoRA+ ratio 16) and Prodigy 1 (LoRA+ ratio 16). Note that AdamW is overtrained but I was just confirming the LR being a factor of slow performance. 
![Image](https://github.com/user-attachments/assets/5f2c2ac3-6e93-44ed-bf43-c96f1ef2ba57)

I would  love to utilize all the experimental features in this optimizer so just trying to figure out why I am getting poor performance. I can run any number of experiments probably but my training speed is very limited. I can do longer training but ideally when performance improves. 

Thank you!

**Comment by LoganBooker** [comment]
Hey @rockerBOO, thanks for the kind words. The optimiser is built on top of the great work of others, I've just done my best to pull a lot of improvements and ideas into the one place. FYI, I use your LoRA inspector quite a bit; handy to check when I forget which layers I trained in older LoRAs. :P

I haven't used or trained Flux at all, so I don't have any immediate insight into the behaviour you're seeing. This may be a silly question but, is the model still learning? The weight norms you show look pretty healthy for a LoRA -- I've found anything over 4-5 in terms of a max norm is usually starting to overfit (which seems to track with the findings in [this paper](https://www.cs.toronto.edu/~rsalakhu/papers/srivastava14a.pdf) (section "A.3 Max-norm Regularization"), which was the original source for Kohya's scale weight norm feature. Again, this might be completely different for Flux.

If you need to boost the learning rate, the only parameter you need to change is `d_coef`. `use_bias_correction=True`, `use_orthograd=True` and `split_groups_mean=False` will all dampen the growth of `d`, rather help it. And if `use_speed=True` gets a similar LR, then there's something more fundamental going on.

At it's core (and greatly simplifying things), Prodigy increases the LR until the weights start to align with the gradient, which should indicate (at the very least) the lowest LR needed to get the model learning.

What this means is that, even with a higher `d_coef`, if the numerator is collecting negative values (indicating that training is making progress at the current LR), then even a very high `d_coef` won't have a significant impact on the end learning rate (it'll just get there a bit faster, or overshoot, and magnify the negative values to discourage LR increases).

I will try and investigate further, but it might be worth using the original Prodigy (with `slice_p=11` to reduce the VRAM usage) and see what LR it ballparks. If it's in the same magnitude as Prodigy+SF, it could mean Prodigy's approach to calculating the LR isn't a good fit for Flux.

> Side question I could make a new issue but I was curious if it was possible to decouple the schedulefree aspect

It's possible, but it does add additional complexity and maintenance to the optimiser (it's already getting a little unwieldly as it is with all the features). I will certainly look into it and see if I can do it in an elegant way. I've been considering a refactor of sorts -- I'm not a big fan of having 400 different features with their own boolean argument. I'm sure there's a cleaner way of doing that.

**Comment by rockerBOO** [comment]
Thank you for your in-depth response. Gave me a lot to look into as well as what different values may effect this. 

I think this example gives a clear picture of the differences. 

AdamW = Cyan
Prodigy = Yellow
ProdigyPlus = Pink

AdamW 7.5e-4 with Cosine LR scheduler
Prodigy d0=7.5e-4 d_coef=2 with Cosine LR scheduler 
ProdigyPlus+ScheduleFree d0=7.5e-4 d_coef=2 

![Image](https://github.com/user-attachments/assets/03958dee-6cdd-4714-bbe4-b334d0cc209a)

![Image](https://github.com/user-attachments/assets/77a8e307-0160-4279-b11f-78cb6c804750)

I've been doing a lot of tests and I think it is related to the schedulefree aspect of it. Schedulefree seems to do a good job at optimizing to lower the norm values. I did some attempts to break it and it would also correct the norm values to be back within a smaller range. It did end up producing reasonable (but not great) results after going up to 1 loss. 

ProdigyPlus+ScheduleFree (i forget all the parameters but I can look them up
d0 = 1e-4

![Image](https://github.com/user-attachments/assets/99534646-b881-456c-9ae1-095e4475596d)

![Image](https://github.com/user-attachments/assets/d7014922-43a5-47a1-a274-36f2b5b24f22)

![Image](https://github.com/user-attachments/assets/f0794ca2-78ad-4956-bbcd-24ab92e8c014)
![Image](https://github.com/user-attachments/assets/9a6c48af-b5aa-451b-93a9-c66780711fa9)

I will report back with more findings but trying to crack into the Flux training a bit to improve training speed still. 

About using a LR scheduler, no worries about it but was curious if it was possible to utilize, but seems much more involved. 

Glad to hear you're using LoRA inspector too :)

**Comment by LoganBooker** [comment]
Hey rockerBOO, appreciate the additional tests and graphs.

The weight progression isn't too surprising -- it might help to visualise how schedule-free's "schedule" compares to cosine.
![Image](https://github.com/user-attachments/assets/41d40e04-63c6-48d7-af3f-2885ffc9bb70)

Again, this is *greatly* simplifying what's happening -- SF decays the weights, then adds the average of all updates so far, before finally applying a small portion of the current update directly to the weights (so they have an immediate influence on the direction).

This is why SF usually requires a higher LR than straight AdamW, so that the updates continue to meaningfully contribute past the early stages of training.

One thing you can try is disabling StableAdamW (`use_stableadamw=False`). This disables the internal gradient scaling and normalisation, but this might be necessary for something like LoRAs, which are trained in short runs and need large updates over stable ones.

Finally, there is a way to "plug-in" to SF's internal schedule, effectively replacing the 1/t one.

All you'd need to do is replace in `prodigy_plus_schedule_free.py`:
```
        weight = dlr ** 2
        weight_sum = group['weight_sum'] + weight
        ckp1 = weight / weight_sum if weight_sum else 0
```
With:
```
        weight_sum = 0
        ckp1 = group['lr']
```
And in `core_optimiser.py`
```
        dlr = (self.shared_d if self.split_groups and self.shared_d else group['d']) * group['lr']
```
With:
```
        dlr = (self.shared_d if self.split_groups and self.shared_d else group['d'])
```

**Comment by rockerBOO** [comment]
Prodigyplus+schedulefree d0 = 1e-4, d_coef=2    "use_stableadamw=False",  "use_speed=True". I also applied your code changes but not sure it's application to the results. 

![Image](https://github.com/user-attachments/assets/101d61d6-60a2-4ab6-98fd-dcb96621fb6b)

![Image](https://github.com/user-attachments/assets/57b93160-cd48-40f3-ba24-79b3bf13c8e1)

it starts to go up but then it sinks down to like nothing for the norms.

I did some more tests at increasing the d0.

![Image](https://github.com/user-attachments/assets/140bc0a5-b26d-4272-aa9c-804fdc710b0e)

All are prodigyplus+schedulefree except "earthly-breeze" which is the one that goes up and flattens out around 15 norm. That's from prodigy d0=5e-4, d_coef=2 and cosine LR scheduler. 

So I am kinda stuck at making it work well enough for Flux to be similar to prodigy in this example. It feels like i can't get it to be a high enough norm and it always wants to bring it down so I am more or less trying to get it to go high enough and a short run to get it to end in the target range. 

Not sure what else to try at this point though but will try different things to experiment with. 

Thanks!

**Comment by rockerBOO** [comment]
I will try the use_speed off, use_stableadamw=False and not using eps=None. I will also try that code change. 

I am mentioning the norm values, prodigy is producing acceptable norm values but prodigyplus is like reducing the norms. the LR doesn't really move when i set the d0 for either of them (except I am using cosine for prodigy. I need the norms to be higher for the impact to be there for these, from my experimentation. The training can seem to go great but if the norm average is like 5 av norm it's barely impacting it. At 10 avg norm it seems to be doing pretty good but hard to get it there. At 15 avg norm it's probably nearer to the edge of being good or bad but really hard to get prodigyplus to land there. I was showing prodigy was able to get there and stay at that value without the norm collapsing down.

**Comment by rockerBOO** [comment]
So far... that edit seems to have stabilized the problem. The norm is no longer collapsing. Will report back on further results. 

prodigyplus.ProdigyPlusScheduleFree | {'weight_decay': 0.01, 'd_coef': 2, 'd0': 0.00025, 'betas': (0.9, 0.99), 'use_stableadamw': False}       
With LoRA+ ratio 8. 

![Image](https://github.com/user-attachments/assets/d0943807-40e1-42f9-b8df-8647b294df24)

**Comment by rockerBOO** [comment]
I haven't gotten back to this with Flux dev. It was very ridged in the parameters it would work with and not collapse the norm. Otherwise it would also climb and then the norm would get smaller and smaller. Since Flux wants a higher norm (like 10+) it would make the impacts not good enough. With Prodigy and half of CosineAnnealing with the min LR of 1e-6 (so like cosine but ending at 1e-6) and the norm only goes upwards or stabilizes. 

Been using this optimizer with Lumina 2 training and it has worked as expected. 

I'm not quite sure what is causing the norm to collapse in this optimizer.

**Comment by LoganBooker** [comment]
Thanks for following up rockerBOO. Glad to hear it works with Lumina 2. As for Flux dev, it's entirely possible it doesn't like schedule-free's "schedule". I'll have another look at making SF a toggle for the next version, though this will 100% have different training characteristics as the Prodigy updates will also change (as we'll swap SF's momentum for a regular first moment).

**Comment by rockerBOO** [comment]
I did some more testing. I found that it worked a bit better with setting the LR to 10 in most cases, and sometimes this was a little high so I went to 5 LR but most of the time 10 LR was adequate. I am currently testing RAdamScheduleFree and found it has similar behavior where the higher LR seems to work better (same 10x LR of what I would use with AdamW). With longer runs it is still occasionally dropping the norm (goes up and then back down) but doesn't seem to have as much adverse effects to it.

**Comment by LoganBooker** [comment]
Interesting -- the main Schedule-Free repo mentions that LRs 2x-10x than normal might be required, which might be playing out here.

If you'd like to try an experimental version of the optimiser with the ability to toggle SF on and off (so only Prodigy is enabled), you can try it here: https://github.com/LoganBooker/prodigy-plus-schedule-free/releases/tag/v2.0.0-dev2. Fair warning, it also includes quite a few changes to the internals, so it will have naturally different behaviour to the current stable version. It still needs a lot of testing.

**Comment by LoganBooker** [comment]
@rockerBOO Alrighty, here's the current dev version I have with the Schedule-Free toggle: https://github.com/LoganBooker/prodigy-plus-schedule-free/releases/tag/v2.0.0-dev3

As mentioned in the previous comment, the internals are somewhat different to the current stable release, so you may need to tweak your arguments. I'll provide more detailed notice for the full release.

**Comment by rockerBOO** [comment]
@LoganBooker Thank you for this! I am trying it out now. For Kohya training it has a look up for ScheduleFree at the end of the class name to see if it's schedulefree or not. 

So I'm just doing this but something to consider. 

```python
class ProdigyPlus(ProdigyPlusScheduleFree):
    pass
```

I can try to update Kohya to be less diligent on the ScheduleFree but maybe it makes sense to allow this format to work.
